src/main/java/ch/njol/skript/expressions/ExprBannerPatterns.java

```diff
@@ -0,0 +1,236 @@
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.BannerMeta;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Consumer;
+
+@Name("Banner Patterns")
+@Description({
+	"Banner patterns of a banner.",
+	"NOTE: When setting a direct placement of a banner pattern, filler banner patterns will be added up to the provided placement."
+})
+@Examples({
+	"broadcast banner patterns of {_banneritem}",
+	"broadcast 1st banner pattern of block at location(0,0,0)",
+	"clear banner patterns of {_banneritem}"
+})
+@Since("INSERT VERSION")
+public class ExprBannerPatterns extends PropertyExpression<Object, Pattern> {
+
+	static {
+		Skript.registerExpression(ExprBannerPatterns.class, Pattern.class, ExpressionType.PROPERTY,
+			"[the] banner pattern[s] of %itemstacks/itemtypes/slots/blocks%",
+			"[the] %integer%[st|nd|rd] [banner] pattern of %itemstacks/itemtypes/slots/blocks%");
+	}
+
+	private Expression<?> objects;
+	private Expression<Integer> patternNumber;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (matchedPattern == 0) {
+			objects = exprs[0];
+		} else {
+			//noinspection unchecked
+			patternNumber = (Expression<Integer>) exprs[0];
+			objects = exprs[1];
+		}
+		setExpr(objects);
+		return true;
+	}
+
+	@Override
+	protected Pattern @Nullable [] get(Event event, Object[] source) {
+		List<Pattern> patterns = new ArrayList<>();
+		Integer placement = patternNumber != null ? patternNumber.getSingle(event) : null;
+		for (Object object : objects.getArray(event)) {
+			if (object instanceof Block block && block.getState() instanceof Banner banner) {
+				if (placement != null) {
+					if (banner.numberOfPatterns() >= placement)
+						patterns.add(banner.getPattern(placement - 1));
+				} else {
+					patterns.addAll(banner.getPatterns());
+				}
+			} else {
+				ItemStack itemStack = ItemUtils.asItemStack(object);
+				if (itemStack == null || !(itemStack.getItemMeta() instanceof BannerMeta bannerMeta))
+					continue;
+				if (placement != null) {
+					if (bannerMeta.numberOfPatterns() >= placement)
+						patterns.add(bannerMeta.getPattern(placement - 1));
+				} else {
+					patterns.addAll(bannerMeta.getPatterns());
+				}
+			}
+		}
+		return patterns.toArray(new Pattern[0]);
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (patternNumber != null) {
+			if (mode == ChangeMode.SET || mode == ChangeMode.DELETE)
+				return CollectionUtils.array(Pattern.class);
+		} else if (mode == ChangeMode.SET || mode == ChangeMode.REMOVE || mode == ChangeMode.ADD || mode == ChangeMode.DELETE) {
+			return CollectionUtils.array(Pattern[].class);
+		}
+		return null;
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
```

Comment:
this method is over a hundred lines long; it needs to be split up.
the single pattern vs multiple pattern logic should be handled in two separate methods, at least.